linkerd/app/outbound/src/opaq/concrete.rs

```diff
@@ -323,12 +324,27 @@ where
         metrics::OutboundEndpointLabels {
             authority,
             labels: metrics::prefix_labels("dst", self.metadata.labels().iter()),
+            zone_locality: self.param(),
             server_id: self.param(),
             target_addr: self.addr.into(),
         }
     }
 }
 
+impl<T> svc::Param<OutboundZoneLocality> for Endpoint<T> {
+    fn param(&self) -> OutboundZoneLocality {
+        if let Some(is_zone_local) = self.metadata.is_zone_local() {
+            if is_zone_local {
+                OutboundZoneLocality::Local
+            } else {
+                OutboundZoneLocality::Remote
+            }
+        } else {
+            OutboundZoneLocality::Unknown
+        }
+    }
+}
```

Comment:
if feels a little unfortunate that this is repeated with `linkerd/app/outbound/src/http/concrete.rs`. i wonder if we could share some code somehow. if we added a constructor like...

```rust
impl OutboundZoneLocality {
    fn new(metadata: &Metadata) -> Self {
        // if let Some ... 
    }
}
```

we might be able to dry these up, and have each of these `svc::Param<OutboundZoneLocality>` implementations both delegate to that shared logic.

Reply:
Agree, this is one thing in here I wanted to fix before converting to a regular PR. Good call on the constructor, that's made this considerably cleaner.